src/lang/pretty_print.rs

```diff
@@ -0,0 +1,669 @@
+//! Pretty-printing functionality for POD2 custom predicates
+
+use std::fmt::Write;
+
+use hex::ToHex;
+use itertools::Itertools;
+
+use crate::middleware::{
+    CustomPredicate, CustomPredicateBatch, NativePredicate, Predicate, StatementTmpl,
+    StatementTmplArg, TypedValue, Value,
+};
+
+/// Trait for converting AST nodes to Podlang source code
+///
+/// This trait provides a consistent interface for pretty-printing different
+/// types of AST nodes back to their Podlang source representation.
+pub trait PrettyPrint {
+    /// Convert this AST node to a Podlang source string
+    ///
+    /// Uses default formatting with no indentation.
+    fn to_podlang_string(&self) -> String {
+        self.to_podlang_string_with_indent(0)
+    }
+
+    /// Convert this AST node to a Podlang source string with custom indentation
+    fn to_podlang_string_with_indent(&self, indent: usize) -> String;
+}
+
+impl PrettyPrint for CustomPredicate {
+    fn to_podlang_string_with_indent(&self, indent: usize) -> String {
+        format_predicate_definition(self, indent, None)
+    }
+}
+
+impl PrettyPrint for StatementTmpl {
+    fn to_podlang_string_with_indent(&self, _indent: usize) -> String {
+        self.to_podlang_string_with_batch_context(None)
+    }
+}
+
+impl StatementTmpl {
+    fn to_podlang_string_with_batch_context(
+        &self,
+        batch_context: Option<&CustomPredicateBatch>,
+    ) -> String {
+        let mut result = String::new();
+
+        match &self.pred {
+            Predicate::Native(native_pred) => {
+                let _ = write!(&mut result, "{}", format_native_predicate(native_pred));
```

Comment:
I guess this `let _ = ` is to capture the `Result` because otherwise clippy complains?
And this `write!` will always return Ok because the writer is a `&mut String`, and a string will never return error on a write.

Reply:
This is eliminated due to the writer change.